i6_models/parts/best_rq/quantizer.py

```diff
@@ -0,0 +1,37 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.linalg import vector_norm
+
+__all__ = [
+    "RandomProjectionQuantizer",
+]
+
+
+class RandomProjectionQuantizer(nn.Module):
+    """
+    implement the fixed random projection quantizer from BestRQ
+    C.f. https://arxiv.org/pdf/2202.01855 for theoretic background
+    code adapted from https://github.com/speechbrain/speechbrain/blob/16b6420d4ff23210cfca2e888be8853264e0cb17/speechbrain/nnet/quantisers.py#L127
+    """
+
+    def __init__(self, input_dim, codebook_dim, codebook_num_vars):
+        """
+        :param input_dim: number of feature dimension of input
+        :param codebook_dim: number of dimension for vocab in the codebook
+        :param codebook_num_vars: vocab size of the codebook
+        """
+        super().__init__()
+
+        self.input_dim = input_dim
+
+        # projection matrix use Xavier initialization
+        P_init = torch.empty((input_dim, codebook_dim))
+        self.register_buffer("P", nn.init.xavier_uniform_(P_init))
+
+        # normalize random matrix for codebook
+        self.register_buffer("CB", F.normalize(torch.randn(codebook_num_vars, codebook_dim)))
+
+    def forward(self, x: torch.tensor) -> torch.tensor:
+        x = F.normalize(x @ self.P)
+        return vector_norm((self.CB.unsqueeze(1) - x.unsqueeze(1)), dim=-1).argmin(dim=1)
```

Comment:
I am wondering about the dimensions here:

`x` starts out as `[B, T, input_dim]` and is then transformed to `[B, T, codebook_dim]` (?)
`CB` has `[codebook_num_vars, codebook_dim]`.

Then should the unsqueezing not be like `self.CB.unsqueeze(0).unsqueeze(0)` to add `B,T` and `x.unsqueeze(1)` to add `codebook_num_vars` so that both have `[B, T, codebook_num_vars, codebook_dim]`.

And then we would do `vector_norm([...], dim=-1)` to get `[B, T, codebook_num_vars]` (with the values being the respective norms) 
followed by `[...].argmin(dim=-1)` which yields `[B, T]` (with the values being the index in the codebook `CB` that best matches the projected `x`.

And then here the indices of the best match are returned.. is this intended? Without the knowledge of the codebook the indices seem worthless to me. Should the returned value not rather be value of the codebook at the specified index?
I should read the paper...
[a few minutes later]
I read the paper: the approach indeed uses the indices of the codebook as targets for a classifier. 
So this part is correct.